### ISSUE TITLE
Mesh Instances randomly not rendering (regression since 3.3)

### ISSUE THREAD

**Issue by tinmanjuggernaut** [godotengine/godot#53374]
### Godot version

3.3.4 official

### System information

Win 10/64 NVIDIA GeForce GTX 1060/PCIe/SSE2

### Issue description

Too frequently some meshes won't render in the editor, and sometimes in game. Switching to another scene and switching back, or closing and reopening will temporarily fix it.

We were developing on 3.2.1 and I never had the issue. As we explored the 3.3 branch I started noticing this error. It's frequent enough that it's a problem. It could have happened any time from 3.2.2. 

@Arnklit Already reported this in #47051, and provided a test project, but it was closed prematurely. Though since it's intermittent I doubt the core devs will see it until they start making a large scene with tons of objects. @mbrlabs also confirmed the problem in that ticket. 

It happens to all sorts of meshes. It has also affected @Zylann's heightmap terrain, one time, in the editor. One quad (say 5mx5m) was invisible and I couldn't get it to become visible until reloading the scene (prior to any of my gdnative work).

Here I'm building a house from a kit. I already had this scene open. I was in another scene, then switched back to it and saw this. I moved the camera closer for the shot:

![image](https://user-images.githubusercontent.com/632766/135769457-b07b2c57-644b-432b-8a4d-692712339c6c.png)

I switched to the other scene, then came back, and the two meshes rendered again:

![image](https://user-images.githubusercontent.com/632766/135769470-cdd2ba16-a990-4239-aad4-3668e910f537.png)

The floor is a fairly simple mesh with one wood material. The wall is more complex. Notice in the first picture you can see the shutter and glass. Here is the full object:

![image](https://user-images.githubusercontent.com/632766/135769574-01408cc2-4176-4188-9558-49fe12007d17.png)

This wall object has separate glass and shutter meshes. The mesh that disappeared includes two materials, the same wood used on all objects in all of these pictures, and a rusty metal frame. Both materials on the mesh didn't render, and both materials are visible on all the other meshes.

The mesh instances have their mesh data stored in the `tscn` file. They were originally imported from fbx, but that file has already been deleted. The textures are separate dds files. The materials are separate binary files, though most of our other materials are `tres`. Here we are using MTLOD and an immediate geometry node, however I've experienced this on our meshes that are not using MTLOD. These shots are all close enough that we're only looking at LOD0. 

I do not believe this is not an issue with LOD, file types or materials. It decides whether it is going to render these objects when the scene opens. Nothing I can do in the scene will get it to render. Not selecting it, moving it, changing visibility, or moving the camera. Only by switching or reopening scenes fixes (or breaks) these meshes.

My console is full of these errors, which repeat when I switch back and forth between my scenes sometimes. 
```
ERROR: get: Condition "!id_map.has(p_rid.get_data())" is true. Returned: __null
   At: ./core/rid.h:150
ERROR: instance_set_transform: Condition "!instance" is true.
   At: servers/visual/visual_server_scene.cpp:713
ERROR: get: Condition "!id_map.has(p_rid.get_data())" is true. Returned: __null
   At: ./core/rid.h:150
ERROR: instance_set_scenario: Condition "!instance" is true.
   At: servers/visual/visual_server_scene.cpp:629
ERROR: get: Condition "!id_map.has(p_rid.get_data())" is true. Returned: __null
   At: ./core/rid.h:150
ERROR: instance_set_visible: Condition "!instance" is true.
   At: servers/visual/visual_server_scene.cpp:780
ERROR: get: Condition "!id_map.has(p_rid.get_data())" is true. Returned: __null
   At: ./core/rid.h:150
ERROR: instance_attach_skeleton: Condition "!instance" is true.
   At: servers/visual/visual_server_scene.cpp:913
```

### Steps to reproduce

Build out a complex scene for a game with a lot of objects to greatly increase your likelihood of seeing the error.
Now at least 3 known gamedevs experience this problem.

### Minimal reproduction project

Look in https://github.com/godotengine/godot/issues/47051

**Comment by Calinou** [comment]
Try disabling the BVH in the project settings and see if the issue persists.

**Comment by Zylann** [comment]
Last time I experienced meshes not rendering with RID errors was in https://github.com/godotengine/godot/issues/44642, although you might be hitting another variant of that if you also get it in game.

**Comment by lawnjelly** [comment]
> Try disabling the BVH in the project settings and see if the issue persists.

It's worth a try (just in case there is some indirect effect), but these errors are coming from _setting_ object data, not reading, and the BVH doesn't set anything on objects (or read if I remember correctly, all the data is sent in exactly the same way as octree).

I'll see if I can reproduce it.

**Comment by Arnklit** [comment]
Yeah I've been seeing this exact same issue on the last two complicated 3D scenes I've made for my "How Good Can I Make Godot Look" series. Sometimes meshes are just gone until you reopen. But yeah it was so random that I thought it was difficult to log a bug on it.

**Comment by akien-mga** [comment]
Likely the same issue as #50630, and I remember another one with GridMap tiles randomly disappearing but GitHub search is failing me.

**Comment by akien-mga** [comment]
For people who are able to reproduce the issue, it would be useful to test and track down which build first introduced the bug.

According to #47051 it dates back at least to 3.2.4 RC 5, and I haven't seen it reported against 3.2.3. So it would be good to confirm that 3.2.3 is not affected, and which 3.2.4 beta or RC build introduced the regression.
https://downloads.tuxfamily.org/godotengine/3.2.4/

**Comment by mbrlabs** [comment]
Yes i remember that issue. I did a lot of testing back then but could not reproduce it reliably (only happend in the editor for me so it was not that critical). I still occasionally have this problem when working on big scenes.

The earliest version i encountered this with was 3.2.4 beta3. See https://github.com/godotengine/godot/issues/47051#issuecomment-802819194

**Comment by akien-mga** [comment]
So far it sounds like you might all be on Windows with Nvidia drivers? Not that it's necessarily a factor as that's the most common gamer config ;)

**Comment by mbrlabs** [comment]
Yes, i'm on Windows 10 with the latest Nvidia drivers.

**Comment by tinmanjuggernaut** [comment]
I will experiment with the bvh settings and other versions and see if I can narrow it down. It might take some time. A change in the bvh was exactly the kind of thing I was looking for, so I hopeful this is it. 

I was working on importing the village kit and switching between scenes hundreds of times over a few hours yesterday. I noticed this rendering bug occur twice in all that time. It likely has hidden meshes that weren't obvious. But it still happens consistently, maybe daily. 

The console errors may or may not be related.

I'm using the nvdia studio driver. The issue has been going on so long that I've upgraded my drivers at least a few times.

**Comment by lawnjelly** [comment]
> I'm using the nvdia studio driver. The issue has been going on so long that I've upgraded my drivers at least a few times.

I suspect the relationship with drivers may be a red herring - it is likely this is affecting the timing, and it is likely a semi-random timing issue which is triggering the bug. Driver issues can cause objects to disappear, but I struggle to think of how they could cause error message our side with RIDs etc.

**Comment by tinmanjuggernaut** [comment]
Would the BVH apply to lights? Here is 3.3.4 w/ `use_bvh` disabled in the project settings. I noticed this issue yesterday and occasionally other times also.

Simple scene, no light is rendered. 

![image](https://user-images.githubusercontent.com/632766/135918721-e3bc61c2-f2ff-498c-87e7-f4d1bd60dce1.png)

Switching scenes reports errors (below) but does not fix the light. Closing and reopening it once did not fix it. A second time it finally loaded.

![image](https://user-images.githubusercontent.com/632766/135918837-20c39e7c-5adb-4cff-98f5-68c9d5b77253.png)

My other asset scenes I'm making with the same set up render the lighting just fine.


Or what about gizmos? Sometimes they disappear as well. Unselecting and re-selecting the collision shape, no gizmo. 

![image](https://user-images.githubusercontent.com/632766/135923654-f5f5d84d-fafb-4bc1-a4e2-5a5ea5fff026.png)

Switch to another loaded scene and back, select the object:

![image](https://user-images.githubusercontent.com/632766/135923727-63d69809-3a14-42b4-b059-11f9f59f9a72.png)


Errors w/ a few additional line numbers:
```
ERROR: instance_set_visible: Condition "!instance" is true.
   At: servers/visual/visual_server_scene.cpp:780
ERROR: get: Condition "!id_map.has(p_rid.get_data())" is true. Returned: __null
   At: ./core/rid.h:150
ERROR: instance_set_scenario: Condition "!instance" is true.
   At: servers/visual/visual_server_scene.cpp:629
ERROR: get: Condition "!id_map.has(p_rid.get_data())" is true. Returned: __null
   At: ./core/rid.h:150
ERROR: instance_set_transform: Condition "!instance" is true.
   At: servers/visual/visual_server_scene.cpp:713
ERROR: get: Condition "!id_map.has(p_rid.get_data())" is true. Returned: __null
   At: ./core/rid.h:150
ERROR: instance_attach_skeleton: Condition "!instance" is true.
   At: servers/visual/visual_server_scene.cpp:913
ERROR: get: Condition "!id_map.has(p_rid.get_data())" is true. Returned: __null
   At: ./core/rid.h:150
ERROR: instance_set_base: Condition "!instance" is true.
   At: servers/visual/visual_server_scene.cpp:456
ERROR: get: Condition "!id_map.has(p_rid.get_data())" is true. Returned: __null
   At: ./core/rid.h:150
```

**Comment by lawnjelly** [comment]
> Would the BVH apply to lights? Here is 3.3.4 w/ `use_bvh` disabled in the project settings. I noticed this issue yesterday and occasionally other times also.

Yes it also applies to lights, BVH is the spatial partitioning scheme used for culling. However, if these screenshots etc are still showing problems with the BVH disabled in `rendering/quality/spatial_partitioning/use_bvh`, then that would seem to rule it out as the direct cause of the problems (because the bug is still occurring when you are using octree instead).

I stress _direct_ because it is also possible that the problem is occurring due to misuse of the spatial partitioning scheme by the client code. I.e. if the client code is creating the same objects multiple times in the tree or not deleting them correctly, the culling can still potentially return this deleted object as being seen by a camera.

One possibility is that these bugs are being caused by multithreaded access to the spatial partitioning scheme, by it octree or bvh. This is not meant to be supported, but to try and identify this as a cause, I have added the project setting `rendering/threads/thread_safe_bvh`. If you enable this and enable the BVH, and the problem disappears, then it would identify the problem as being caused by multithread access.

There are also lots of other possibilities, this may have nothing to do with spatial partitioning. The reason for the speculation is that the BVH was introduced around the time this bug was first reported, but it may be coincidence.

**Comment by tinmanjuggernaut** [comment]
It just happened with hterrain. Twice in one day with threadsafe bvh turned on. :/

![image](https://user-images.githubusercontent.com/632766/137635917-e951921f-7321-4035-b3d4-4cc86328b918.png)

**Comment by tinmanjuggernaut** [comment]
> Mesh Instances randomly not rendering in the editor (regression since 3.3)

@akien-mga though almost all of the instances I have experienced have been in the editor, I have had meshes not appear while running in game since 3.2.1. And since the editor is a running game, there's really little difference between the renderer and visual server for either scenario.

**Comment by tinmanjuggernaut** [comment]
I tried running my game in 3.2.1, but there are too many problems with it. Code that should work is invalid. It doesn't read all of the meshes that were imported in 3.3 or 3.4 and saved in a native mesh format. So I'm not going to work backward.

Instead, I built a new 3.x da5c843bd debug version w/ threadsafe BVH on, and worked today while running the editor through the MSVC debugger. I set some breakpoints in the code and recorded a video of my screen whenever I hit an error to debug the callstack and objects.

Here are some conclusions so far:

1. The errors are happening all the time. While opening, closing, and creating object scenes, and switching between scenes, I'm hitting a block of 3-4 errors at least hourly if not more like every 10-15 minutes. However the times I hit an error that provides an obvious visual artifact is much rarer, maybe 1/10th the time.
2. The problems occur when VisualServer wants to `set_visible`, `set_transform`, or `set_scenario` on a particular resource (RID), which happens on all objects every time it switches scenes, including editor viewport objects (e.g. SpatialEditorViewport whatever that is). Only once in a while does it trigger on `set_visible`, hence point 1. On each scene switch, the scene goes through `_enter_tree()` and `_exit_tree()`. 
3. The Set (red-black tree) that stores the RIDs of the objects in the scene is not properly populated. These errors are occurring because the RIDs of the objects in the scene are missing from the Set of RIDs the visual server queries.

Here is one of the three areas where the error messages come from, where VisualServer attempts to get the instance of the RID.

```
VisualServerScene::instance_set_transform(RID p_instance, const Transform &p_transform) {
	Instance *instance = instance_owner.get(p_instance);
	ERR_FAIL_COND(!instance);
```

`instance_owner` is an `RID_Owner`, which is a class with a `Set` called id_map. A Set is a red-black tree and is used to contain all RIDs for the scene: `Set<RID_Data *> id_map;`

Here is `instance_owner.get()`
```
rid.h::get(p_rid)
		ERR_FAIL_COND_V(!p_rid.is_valid(), nullptr);
		ERR_FAIL_COND_V(!id_map.has(p_rid.get_data()), nullptr);
```

The first line doesn't produce an error, so the RID object is valid.

I expanded the second macro and added some debugging code to determine that in almost all cases the RID object and the resource ID integer within are valid. However when this fails, the ID is NOT in the Set (id_map). In only one instance today was the RID data invalid.

4. I dumped the entire red-black tree across several error instances to confirm that the element wasn't just misplaced. It's unlikely the problem is in the implementation of the tree. It's much more likely that the problem occurs when the Set is being populated. On another day I'll explore that.

Here's a video showing my debugging progress. It's long, and gets more efficient in later chapters. Look for the chapter markers on the timeline. I captured 5 error sessions, but only one was visual, at 1:06:45.

https://youtu.be/ozET9itjeD4

**Comment by tinmanjuggernaut** [comment]
Are the octree or bvh related to populating `id_map` or is that entirely
unrelated and I can ignore them?




>

**Comment by lawnjelly** [comment]
> Are the octree or bvh related to populating `id_map` or is that entirely unrelated and I can ignore them?

I think based on the evidence so far I don't think the bug is anything to do with octree or bvh, that was most likely a red herring.

**Comment by mbrlabs** [comment]
I encountered this again today in the editor with a custom 3.4rc1 build. The error message was slightly different however:

```
 C:\Users\mbr\Projects\godot3\core/rid.h:149 - Condition "!id_map.has(p_rid.get_data())" is true. Returned: nullptr
 drivers\gles2\rasterizer_storage_gles2.cpp:2724 - Condition "!mesh" is true. Returned: VisualServer::PRIMITIVE_MAX
 servers\visual\visual_server_scene.cpp:536 - Condition "instance->base_type == VisualServer::INSTANCE_NONE" is true.
```

**Comment by tinmanjuggernaut** [comment]
@mbrlabs The core issue is this part of all of the various error messages: `!id_map.has(p_rid.get_data()) is true`. This is because the RID was previously removed from its container (`id_map`) for an unknown reason.

@lawnjelly 
I can't use the callstack PR as it requires EngineDebugger which doesn't exist in 3.x. However, I learned how to dump the call stack with CaptureStackBackTrace() in the Windows SDK.

I ran through several errors and got 4 unique traces ([callstacks.xlsx](https://github.com/godotengine/godot/files/7464546/callstacks.xlsx)). All of them go back to `VisualServerWrapMT::free` (as a wrapper for `VisualServerRaster::free`), then diverge from there. 


One thing I have found is that `VisualServerRaster::free()` is often called with invalid RIDs (ie `! p_rid.is_valid()` and sometimes with RID==0, which should never happen (Since `RID_OwnerBase::init_rid()` guarantees it will always be >=1). These invalid conditions are then sent to 5 VisualServers to be searched for in 5-10 redblack trees each. I'll introduce some sanity checks to skip these conditions.

When running in a fully loaded scene, not switching, free() is called often, including both valid and invalid calls. On my village gallery scene shown above, it's called around 10 times per frame, whenever the camera or an object is moved. 

I haven't yet figured out why these objects are being freed or if they are legitimate, why only 10 free calls are being made when there are 300 objects in this scene, or why invalid RIDs and RID==0s are attempting to be freed over and over again.

So far, searching backwards with the callstacks hasn't been effective due to the design. I've started tracing forward by logging markers at various points in the code, starting with what happens when editor tabs are switched. So far on one instance, I got errors when this line adds the new scene to the tree: 
```
EditorNode::set_current_scene()
     scene_root->add_child(new_scene);
```
I haven't traced into add_child() yet.

**Comment by Zylann** [comment]
Just as a reminder, #44642 identifies a place in the `SpatialEditor` that frees rids that dont belong to the editor. With the checks you are adding, you could verify this place too to make sure it is not the same cause.

**Comment by tinmanjuggernaut** [comment]
Thanks. I didn't know about that one. It certainly looks suspicious, and that class shows up in some of my call stack traces. I was debugging it quite a bit today, but nothing conclusive yet. 

Also I just realized where I pointed out that errors were occurring (add_child) is only when the problem discovered and messages are printed. The RIDs are probably being freed before then, and my logs should be able to identify that.

**Comment by tinmanjuggernaut** [comment]
Turning off the grid in the viewport prevents the error messages and symptoms on my system. Specifically disabling all three `Editor->Editor Settings->Editors->3d->grid_*_plane`, though only `xz` is enabled by default. 

Guys, please test this on your systems and let me know if you get any `!id_map.has(p_rid.get_data())` error messages. You should be able to test this with any build of 3.3 and later. 

If that fixes it for everyone, then it looks like I've solved the issue. 

### The Cause

SpatialEditor has two 3-item arrays: `grid` and `grid_instance`. These hold the meshes and instances of the 3D grid shown in the viewports for each of the 3 grid planes. 

These arrays are continually freed and repopulated whenever the camera or an object is moved, or several other events. They are repopulated only if the grid is enabled. Which means that sometimes (all the time), the old RIDs are repeatedly freed.

**This is the root problem: It is not safe to free RIDs multiple times. When they are freed, they leave a dangling pointer.**

Every update cycle in SpatialEditor, RIDs are sent to the visual server to be freed. Sometimes with current pointers, sometimes with dangling pointers.

VisualServer has no way to tell if an RID is valid or not. `rid.is_valid()` is true even if its dangling, and `rid.get_id()!=0)` will cause an exception if the pointer is dangling. 

VS gets the pointer from each RID and searches for it in each of ~20 red black trees to see if the pointer matches any instance in any tree. If there's no match, nothing happens. However if there is a match, the instance is removed from the tree.

When you have a scene with hundreds of objects, it only takes a few scene switches to generate thousands of compares. Eventually it finds a match and a random instance gets removed from whichever r/b tree owned it (cameras, lights, meshes, etc).

The instance is still referenced elsewhere in the engine, but since the VisualServer can't find the instance in any of its id_maps, it spams the console with error messages, usually for the same instance over and over. Switching scenes or closing/reopening scenes causes SpatialEditor to repopulate its grid arrays, but it could free another random instance.

### The Fix
I could just edit SpatialEditor and manually assign RID() to the grid arrays after free(). However this issue has highlighted a bigger problem in the engine. Freeing RIDs leaves dangling pointers. It's not safe to free an RID multiple times. If you do, you could remove instances from random id_maps throughout the engine, not just in VisualServer but also in Physics and anywhere else they are used.

My PR will tweak `RID` and every instance of `free(RID)` to enforce cleaning up these dangling pointers (passing the RIDs by reference and setting `_data=nullptr`). This also means `RID.is_valid()` will actually mean something. I've already got it working in all the VisualServers. I just need to do it for Physics and any other place. I'll push up a PR probably tomorrow with the RID changes and VS::free() sanity checks.

Also, RID.h, handles release and debug mode differently. Debug mode searches a red/black tree to see if it contains an RID. In release mode it doesn't even use a red/black tree. Instead the RID just keeps a direct reference to its owner. I really don't like having two versions of code, with release mode having *more* functions available. I intend to bring the two versions closer together to reduce complexity of finding bugs in the future.

**Comment by lawnjelly** [comment]
@tinmanjuggernaut well done for finding this! Yes that does sound plausible.

Before spending any time attempting changes to RID I'd recommend discussing with reduz (on godot rocket chat), as it is a very core area of the engine (he may want to make any changes there himself, or propose a different solution).

**Comment by Zylann** [comment]
> My PR will tweak RID and every instance of free(RID) to enforce cleaning up these dangling pointers (passing the RIDs by reference and setting _data=nullptr)

You cannot have this work for the script API, `free` is exposed and RIDs cannot be passed by reference there... so your best option still remains to reset every freed RID.

**Comment by tinmanjuggernaut** [comment]
I had to tear RID apart to find the issue, so I'm familiar with it. I will ask reduz.

Either every caller of any `free(RID)` needs to also do `rid=RID()`, or RID needs to clean itself up, or a combination, like it's clean in the engine and dirty in script. 

I currently have it working the original way for the script API, and a clean way for the engine. Several areas of the engine regularly send bad data to free.

As for script, it could possibly generate an error if desired, but there's no way for it to know that the pointer is dangling, unless 1) it try{}s to access the memory and 2) it's not pointing to a valid object, but this issue demonstrates that it does far too often. Many times I watched `SpatialEditor`, which should have only referenced `VisualInstance`, referencing `AnimationNodeOneShot`.

**Comment by tinmanjuggernaut** [comment]
Find the PR at https://github.com/godotengine/godot/pull/54650.
Juan asked @RandomShaper to be the primary reviewer, but I invite all of you guys to look and comment.

**Comment by RandomShaper** [comment]
It would be awesome to get an MRP based in the project in the original post. The one in #47051 has indeed a problem related to tab switching, but it's no longer the one reported there.

What happens there now is related to timing. If you switch to and away from the fur ball scene, a piece of script in the Fur addon causes errors about certain nodes not being in the tree:

```
func update_physics_object(delay : float) -> void:
	yield(_shell_fur_object.get_tree().create_timer(delay), "timeout")
	_physics_pos = _current_physics_object().global_transform.origin
	_physics_rot = _current_physics_object().global_transform.basis.get_rotation_quat()
```

While the fur scene is the active one, `update_physics_object()` is called. If you switch away from it in a shorter time than `delay` (which seems to be 0.5), by the time the `yield()` ends, the "current physics object" is no longer in the tree, causing the errors. I woulnd't call that a Godot bug. It may be possible to freeze the state of a scene in a tab that becomes inactive and unfreeze when it's active again, but it's not possible worh the hassle. Sounds like a not trivial at all thing to implement.

**Comment by akien-mga** [comment]
Fixed by #54650.
